### ISSUE TITLE
aot: Directive inheritance cause `Cannot determine the module` error

### ISSUE THREAD

**Issue by tinayuangao** [angular/angular#16637]
**I'm submitting a ...**  (check one with "x")
```
[x ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Have a Directive `MdCheckboxRequiredValidator` that `extends CheckboxRequiredValidator` from `angular/forms`.
The AOT has error ```Cannot determine the module for class MdCheckboxRequiredValidator in /home/travis/build/angular/material2/dist/packages/demo-app/material/typings/checkbox/checkbox.d.ts! Add MdCheckboxRequiredValidator to the NgModule to fix it.``` 
But `MdCheckboxRequiredValidator` is already added to NgModule in `src/lib/checkbox/index.ts`
After removed the inheritance, the AOT compiles without error.

**Expected behavior**
No error

**Minimal reproduction of the problem with instructions**
https://github.com/angular/material2/pull/4395
https://travis-ci.org/angular/material2/jobs/230064418

* **Node (for AoT issues):** `node --version` =   
6.9.5

**Comment by tbosch** [comment]
Reproduce locally:

1. clone the branch of this PR
2. `gulp aot:build`

**Comment by chuckjaz** [comment]
@jelbourn @tinayuangao I did more investigation on this bug and determined that the directory layout used to build `demo-app` is not laid out in a the way the complier expects causing it to lose track of where to look for the metadata for checkbox `.d.ts` files.

**STEPS TO REPRODUCE CHANGE**
It is now easier to reproduce this bug than the steps above:
1) clone angular/materal2 and npm install
2) Change the class declaration in `checkbox-required-vaidator.ts` to be: `export class MatCheckboxRequiredValidator extends CheckboxRequiredValidator {}`
3) `gulp aot:build`

**ISSUE**

The `flatModuleRedirect` handling has an assumption that the metadata will redirect to a sub-directory of the final metadata instead of the same directory as is the case for `checkbox.d.ts`. The compiler marks the directory holding the index as being a `sources files` directory and will not recheck the state for `.d.ts` files in that directory. This causes `checkbox-required-validator.d.ts` to be treated as a source file and collects metadata from `checkbox-required-validator.d.ts` instead of using the metadata in `index.metadata.json`. This  is reported as an issue here, and not elsewhere, because `MatCheckboxRequiredValidator` inherits from `CheckboxRequiredValidator` which we find the correct metadata and this class is marked as a directive but this symbol is not in a module ( its bundled alias is).

The current work around works because the compiler will ignore classes that have expression in the `extends` clause of a class. This work-around will break if:

1) tries to support expression by simplifying the expression to the class
2) `CheckboxRequiredValidator` changes its metadata that is copied in `MatCheckboxRequiredValidator` or adds metadata that is required that `MatCheckboxRequiredValidtor` doesn't copy.

**POSSIBLE SOLUTIONS**

1) Change the directory structure of `@angular/material/checkbox` to place `checkbox.ts` and `checkbox-required-validator.ts` in subdirectory of `lib/checkbox`. This has the advantage of working for 4.4 as well as 5.0.

2) Fix 5.0 to handle files in the same sub-directory as the index or improve the detection of bundled `.d.ts` files to use the `origins` to determine which `.d.ts` files were bundled.

**Comment by JoostK** [comment]
The metadata system is no longer used in Angular 9.0 with Ivy, so this issue has been resolved.
